craft_application/services/service_factory.py

```diff
@@ -62,6 +91,68 @@ class ServiceFactory:
 
     def __post_init__(self) -> None:
         self._service_kwargs: dict[str, dict[str, Any]] = {}
+        self._services: dict[str, services.AppService] = {}
+
+        factory_dict = dataclasses.asdict(self)
+        for cls_name, value in factory_dict.items():
+            if cls_name.endswith("Class"):
+                if value is not None:
+                    identifier = _CAMEL_TO_PYTHON_CASE_REGEX.sub(
+                        "_", cls_name[:-5]
+                    ).lower()
+                    warnings.warn(
+                        f'Registering services on service factory instantiation is deprecated. Use ServiceFactory.register("{identifier}", {value.__name__}) instead.',
+                        category=DeprecationWarning,
+                        stacklevel=3,
+                    )
+                    self.register(identifier, value)
+                setattr(self, cls_name, self.get_class(cls_name))
+
+        if "package" not in self._service_classes:
+            raise TypeError(
+                "A PackageService must be registered before creating the ServiceFactory."
+            )
+
+    @classmethod
+    def register(
+        cls,
+        name: str,
+        service_class: type[services.AppService] | str,
+        *,
+        module: str | None = None,
+    ) -> None:
+        """Register a service class with a given name.
+
+        :param name: the name to call the service class.
+        :param service_class: either a service class or a string that names the service
+            class.
+        :param module: If service_class is a string, the module from which to import
+            the service class.
+        """
+        if isinstance(service_class, str):
+            if module is None:
+                raise KeyError("Must set module if service_class is set by name.")
+            cls._service_classes[name] = (module, service_class)
+        else:
+            if module is not None:
+                raise KeyError(
+                    "Must not set module if service_class is passed by value."
+                )
+            cls._service_classes[name] = service_class
```

Comment:
I may asked about it before, but should we warn if someone is shadowing already registered service ? 
I could find it useful to see in logs, if there was something already registered under that name instead of blind override

Reply:
It is expected to override at least some of the services, so definitely not a warning. We could log it in `trace`, but that would be a separate feature.